### ISSUE TITLE
Inconsistent dims of the one-hot encoded labels in the CoastTrain dataset

### ISSUE THREAD

**Issue by FlorisCalkoen** [Doodleverse/segmentation_gym#102]
Hi all, many thanks for your great work, but unfortunately I experience some trouble reproducing the results because of an issue with the labels in the training data. 

The "label" or mask dimension is presented a one-hot encoded vectors per px, but I find that these are not consistent among the different images, i.e., dimension of the label is (height, width, range(1,11). This causes the one-hot encoded labels to be inconsistent among multiple images. Please find an example below.  

I downloaded the associated CoastSeg dataset (Wernette et al 2022) from https://cmgds.marine.usgs.gov/data-releases/datarelease/10.5066-P91NP87I/. Specifically I'm looking at the S2 data available at this url:  https://cmgds.marine.usgs.gov/data-releases/media/2022/10.5066-P91NP87I/bdc18f4f38004538af974aa0540d468c/Sentinel2_11_001.zip

To get the data:
```bash
mkdir -p ~/tmp/coasttrain
wget https://cmgds.marine.usgs.gov/data-releases/media/2022/10.5066-P91NP87I/bdc18f4f38004538af974aa0540d468c/Sentinel2_11_001.zip -O ~/tmp/coasttrain/data.zip
cd ~/tmp/coasttrain
unzip data.zip
rm data.zip
```

```python
import pathlib
import pandas as pd
import numpy as np

coasttrain_dir = pathlib.Path.home().joinpath("tmp", "coasttrain")
metadata_fp = coasttrain_dir.joinpath("Sentinel2_11_001.csv")

metadata = pd.read_csv(metadata_fp)

# take random sample fp from df
fp = coasttrain_dir.joinpath(
    metadata.sample(1)["images"].iloc[0].split("/")[1].rsplit(".", 1)[0] + ".npz"
)

# load data into python
data = np.load(fp)
# (height, width, label) - run this a few times to see the diffs in label dimension
print(data["label"].shape)  # e.g., (320, 184, 7) or (750, 956, 8) or (468, 134, 5)
```

In this case, 7, 8, and 5 refer to the length of the one-hot encoded label dimension. 

Any chance you have an updated dataset with matching labels? Or how do you get the labels for the training data?

**Comment by ebgoldstein** [comment]
Hi @FlorisCalkoen ,
thanks for providing a code snippet here, I was able to exactly reproduce your issue. 

You are correct, the one-hot labels change size - this occurs because each image does not neccesary have all classes that a labeler could use (11 in this case). If you look at the `.csv` file in the `.zip` file, you will see some helpful columns for checking this:
- `classes_array` and `num_classes` are the names and the number of classes **available** to the annotator for the image
- `classes_present_array` and `classes_present_integer` lists the name and number of classes **used** by the annotator for that particular image. 

The fact that one-hot labels change size is just a result of this..  

We note in the CoastTrain data summary (on the landing page) that the easiest way to wrangle the data into ML format is to use Doodler, and i recommend this approach. Thsi will get all the labels correctly encoded.. I just verified that it works. 

So, can you follow this workflow and see how it goes:

- Set up [doodler](https://github.com/Doodleverse/dash_doodler) on your machine
- activate the doodler conda environment
- navigate back into the doodler directory
- navigate into `/utils/`
- run `gen_images_and_labels.py`, and select the folder of Coast train .npzs  you want to process (`~/tmp/CoastTrain`)
- Once the script is done, you should see several new folders in `~/tmp/CoastTrain`. Two will be 'images' and 'labels', and they will be named correctly and the labels will be correclty encoded. You can look at the Overlays folder to check it out.
- copy/paste those two folders into your Gym directory, and you will be good to go..

this is an 'official' way to get images and labels from Doodler (CoastTrain is from Doodler) -> gym..

https://doodleverse.github.io/dash_doodler/docs/tutorial-extras/next-steps

some examples:
![chunk1_20190102T184759_20190102T184755_T11SKU TCI_RGB_site47_ID3_overlay](https://user-images.githubusercontent.com/5330599/198601132-1ebc3e77-4771-4840-8c5f-8c092ccdbfd6.png)

![chunk1_20190810T165849_20190810T170927_T14RPP TCI_RGB_site54_ID3_overlay](https://user-images.githubusercontent.com/5330599/198601158-749f4a6f-8e41-4394-aec5-f48ed22df75d.png)

![chunk1_20181215T183751_20181215T184316_T11SKT TCI_RGB_site49_ID3_overlay](https://user-images.githubusercontent.com/5330599/198601110-83915cdd-92ae-47d6-8f5e-6e74d59806a9.png)

**Comment by FlorisCalkoen** [comment]
Ok, thanks, for suggesting that workflow! But.. unfortunately I now have an issue with pydensecrf.. Have you experienced that earlier? I'll try to take some useful snippets from that `gen_images_and_labels.py`. 

Any chance you also have a snippet to associate the geographic coordinates from the metadata to the the numpy arrays? I.e., numpy arrays + metadata to rioxarray dataset. 

The package issue should probably be a seperate issue on the other page, but for sake of completeness, please find some infor below. The package doesn't seem to be available on channel.. I tried to install it in several ways:

```bash
conda create --name dashdoodler python=3.8
conda install -c conda-forge pydensecrf cairo cairosvg scikit-learn scikit-image psutil dash flask-caching requests pandas matplotlib ipython tqdm  # PackagesNotFoundError: The following packages are not available from current channels:
  - pydensecrf
```

```bash
conda create --name dashdoodler pydensecrf # PackagesNotFoundError: The following packages are not available from current channels:
  - pydensecrf
```

```bash
conda create --name dashdoodler python=3.8
pip install pydensecrf
```
results in: 

```
< much output >


× Encountered error while trying to install package.
╰─> pydensecrf

note: This is an issue with the package mentioned above, not pip.
hint: See above for output from the failur
```

**Comment by ebgoldstein** [comment]
Hi @FlorisCalkoen - 
1) i moved the pydensecrf issue to the doodler repository: https://github.com/Doodleverse/dash_doodler/issues/45... Can we continue that discussion in Doodler?

2) Yes I see those 5 examples mixed in with the 340 others:

the `classes array` for those 5:
> 'water', 'whitewater', 'sediment', 'other_bare_natural_terrain', 'marsh_vegetation', 'terrestrial_vegetation', 'agricultural', 'development', 'nodata', 'unusual', 'unknown'

and the typical `classes array` is:
'water', 'whitewater', 'sediment', 'other_natural_terrain', 'vegetated_surface', 'agricultural', 'development', 'cloud', 'nodata', 'unusual', 'unknown'

I think in the short term, yes @FlorisCalkoen , i recommend dropping those rows. 
in the long term, @dbuscombe-usgs what do you think?

(Also, @dbuscombe-usgs , should we transfer this issue to the CoastTrain repository? https://github.com/CoastTrain/CoastTrain)

**Comment by FlorisCalkoen** [comment]
Ok, thanks, I'll keep you posted when I continue with the coasttrain dataset. For now I'll remove that pip install output from my previous comment to keep this issue a bit more readable :)

**Comment by ebgoldstein** [comment]
you can keep it here , its totally fine.. no worries!

**Comment by dbuscombe-usgs** [comment]
@FlorisCalkoen thanks for reporting the issue with the mismatch in labels for those 5 samples. I will take a look at this and see what is going on in the next few days

**Comment by ebgoldstein** [comment]
@dbuscombe-usgs , can you try to 'Transefer' this issue to https://github.com/CoastTrain/CoastTrain? (i would do it, but i don;t have permissions in CT).. If that doesn't work, perhaps we could close this and open a new issue in CT that references this issue? Just to get it out of Gym..

**Comment by dbuscombe-usgs** [comment]
I can only transfer to other repos in this org. I can close and reopen in CT?
